rskj-core/src/test/java/co/rsk/peg/bitcoin/BitcoinUtilsTest.java

```diff
@@ -0,0 +1,294 @@
+
+        Sha256Hash expectedSigHash = btcTx.hashForSignature(
+            0,
+            flyoverP2shErpRedeemScript,
+            BtcTransaction.SigHash.ALL,
+            false
+        );
+
+        // Act
+        Optional<Sha256Hash> sigHash = BitcoinUtils.getFirstInputSigHash(btcTx);
+
+        // Assert
+        Assertions.assertTrue(sigHash.isPresent());
+        Assertions.assertEquals(expectedSigHash, sigHash.get());
+    }
+
+    @Test
+    void test_getFirstInputSigHash_no_input() {
+        // Arrange
+        Address destinationAddress = BitcoinTestUtils.createP2PKHAddress(btcMainnetParams, "destinationAddress");
+
+        BtcTransaction btcTx = new BtcTransaction(btcMainnetParams);
+        btcTx.addOutput(Coin.COIN, destinationAddress);
+
+        // Act
+        Optional<Sha256Hash> sigHash = BitcoinUtils.getFirstInputSigHash(btcTx);
+
+        // Assert
+        Assertions.assertFalse(sigHash.isPresent());
+    }
+
+    @Test
+    void test_getFirstInputSigHash_invalid_input_no_redeem_script() {
+        // Arrange
+        Address destinationAddress = BitcoinTestUtils.createP2PKHAddress(btcMainnetParams, "destinationAddress");
+
+        BtcTransaction btcTx = new BtcTransaction(btcMainnetParams);
+        btcTx.addInput(BitcoinTestUtils.createHash(1), 0, new Script(new byte[]{}));
+        btcTx.addOutput(Coin.COIN, destinationAddress);
+
+        // Act
+        Optional<Sha256Hash> sigHash = BitcoinUtils.getFirstInputSigHash(btcTx);
+
+        // Assert
+        Assertions.assertFalse(sigHash.isPresent());
+    }
+
+    @Test
+    void test_getFirstInputSigHash_many_inputs() {
+        // Arrange
+        List<BtcECKey> signers = Arrays.asList(
+            BtcECKey.fromPrivate(Hex.decode("fa01")),
+            BtcECKey.fromPrivate(Hex.decode("fa02")),
+            BtcECKey.fromPrivate(Hex.decode("fa03"))
+        );
+
+        List<BtcECKey> pubKeys = signers.stream().map(BtcECKey::getPubKeyPoint).map(BtcECKey::fromPublicOnly)
+            .collect(Collectors.toList());
+        signers.sort(BtcECKey.PUBKEY_COMPARATOR);
+
+        int totalSigners = signers.size();
+        int requiredSignatures = totalSigners / 2 + 1;
+
+        Script redeemScript = ScriptBuilder.createRedeemScript(requiredSignatures, signers);
+        Script p2SHScript = ScriptBuilder.createP2SHOutputScript(requiredSignatures, pubKeys);
+
+        Address destinationAddress = BitcoinTestUtils.createP2PKHAddress(btcMainnetParams, "destinationAddress");
+
+        BtcTransaction btcTx = new BtcTransaction(btcMainnetParams);
+
+        for (int i = 1; i <= 50; i++) {
+            btcTx.addInput(BitcoinTestUtils.createHash(1), 0, new Script(new byte[]{}));
+            RedeemData redeemData = RedeemData.of(pubKeys, redeemScript);
+
+            Script inputScript = p2SHScript.createEmptyInputScript(redeemData.keys.get(0), redeemData.redeemScript);
+            btcTx.getInput(i-1).setScriptSig(inputScript);
+        }
```

Comment:
```suggestion
           RedeemData redeemData = RedeemData.of(pubKeys, redeemScript);

            Script inputScript = p2SHScript.createEmptyInputScript(redeemData.keys.get(0), redeemData.redeemScript);
        for (int i = 1; i <= 50; i++) {
            btcTx.addInput(BitcoinTestUtils.createHash(1), 0, new Script(new byte[]{}));
            btcTx.getInput(i - 1).setScriptSig(inputScript);
        }
```

Reply:
I believe each input should have its own instance.

Reply:
The Objects Are Different For Each Loop?

Reply:
Tests were refactored, and the way I assert the sighash is valid was improved. Please check it out.